Add strongly typed representation for `dynamic-range-limit` property

weinig opened the pull request:
#### b3c8e4373f501e022b3e15e152b648f05bcd1802
<pre>
Add strongly typed representation for `dynamic-range-limit` property
<a href="https://bugs.webkit.org/show_bug.cgi?id=287335">https://bugs.webkit.org/show_bug.cgi?id=287335</a>

Reviewed by Tim Nguyen.

Implements parsing, value representation, mix resolution
and animation for &apos;dynamic-range-limit&apos;.

The renderer still needs to do something with the value to make
it useful, so this is behind a new feature flag &quot;CSSDynamicRangeLimitPropertyEnabled&quot;
set to &quot;testable&quot;.

The recursive nature of dynamic-range-limit-mix() means this
representation is a bit more complex than others but only
requires allocation when a mix function is actually used.

* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color-hdr/computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color-hdr/inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color-hdr/interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color-hdr/parsing-expected.txt:
    - Update results to passing.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
* Source/WebCore/css/CSSDynamicRangeLimitValue.cpp: Added.
* Source/WebCore/css/CSSDynamicRangeLimitValue.h: Added.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.h:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/css/values/color/CSSDynamicRangeLimit.cpp: Added.
* Source/WebCore/css/values/color/CSSDynamicRangeLimit.h: Added.
* Source/WebCore/css/values/color/CSSDynamicRangeLimitMix.h: Added.
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/values/color/StyleDynamicRangeLimit.cpp: Added.
* Source/WebCore/style/values/color/StyleDynamicRangeLimit.h: Added.
* Source/WebCore/style/values/color/StyleDynamicRangeLimitMix.cpp: Added.
* Source/WebCore/style/values/color/StyleDynamicRangeLimitMix.h: Added.

Canonical link: <a href="https://commits.webkit.org/290120@main">https://commits.webkit.org/290120@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c946153326bd216f7c4f6724c45d971368b32ed

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89050 "21 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8574 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43779 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94024 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39811 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91101 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8961 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16760 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/68614 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26288 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92052 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/6850 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/80736 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/48977 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6599 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38918 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/81849 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76961 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/36091 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95861 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/87826 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16230 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11871 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/77494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16486 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/76539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/76782 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18920 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21174 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19736 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9322 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16244 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21555 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/110319 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15985 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26480 "Found 4 new JSC stress test failures: wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-slow-memory, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-eager-jettison, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-slow-memory, wasm.yaml/wasm/stress/repro_1289.js.wasm-collect-continuously (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19436 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17766 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -3593,6 +3593,25 @@
+                "high"
+            ],
+            "codegen-properties": {
+                "settings-flag": "cssDynamicRangeLimitPropertyEnabled",
```
Is this sufficient to gate the parsing given the code is not auto-generated? Does `consumeDynamicRangeLimit` also need to check this?

nt1m replied:
Looks like it's being taken care by https://searchfox.org/wubkat/source/__GENERATED__/WebCore/DerivedSources/CSSPropertyParsing.cpp#4802

nt1m commented on Source/WebCore/style/values/color/StyleDynamicRangeLimit.cpp:
```diff
@@ -0,0 +1,112 @@
+    // construct a full fledged CSS::DynamicRangeLimitMixFunction, we can utilize the weighted sum
+    // addition functionality to construct just what we need and then resolve that directly.
+
+    DynamicRangeLimitMixPercentage fromMixPercentage = (1.0 - context.progress) * 100.0;
+    DynamicRangeLimitMixPercentage toMixPercentage   = (context.progress) * 100.0;
+
+    DynamicRangeLimitMixFunction function { .parameters = { 0_css_percentage, 0_css_percentage, 0_css_percentage } };
```
Can `auto` be used here?

weinig replied:
I don’t think so. How are you imagining that?

nt1m commented on Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
Looks like this is missing the corresponding change in https://searchfox.org/wubkat/rev/3cfa20e5d0cd33e53dbc96342d8ac86ef27dad8d/Source/WebCore/rendering/style/StyleRareInheritedData.cpp#401

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-color-hdr/parsing-expected.txt:
```diff
@@ -25,6 +25,6 @@ PASS e.style['dynamic-range-limit'] = "dynamic-range-limit-mix(low, high, 10%)"
 PASS e.style['dynamic-range-limit'] = "dynamic-range-limit-mix(standard, high, 0.1)" should not set the property value
-PASS e.style['dynamic-range-limit'] = "dynamic-range-limit-mix(high 0%, standard 0%)" should not set the property value
-PASS e.style['dynamic-range-limit'] = "dynamic-range-limit-mix(dynamic-range-limit-mix(high 1%, standard 2%) 3%, dynamic-range-limit-mix(constrained-high 0%, high 0%) 6%)" should not set the property value
+FAIL e.style['dynamic-range-limit'] = "dynamic-range-limit-mix(high 0%, standard 0%)" should not set the property value assert_equals: expected "" but got "dynamic-range-limit-mix(high 0%, standard 0%)"
+FAIL e.style['dynamic-range-limit'] = "dynamic-range-limit-mix(dynamic-range-limit-mix(high 1%, standard 2%) 3%, dynamic-range-limit-mix(constrained-high 0%, high 0%) 6%)" should not set the property value assert_equals: expected "" but got "dynamic-range-limit-mix(dynamic-range-limit-mix(high 1%, standard 2%) 3%, dynamic-range-limit-mix(constrained-high 0%, high 0%) 6%)"
```
Do we know why the spec forbids this?

squelart replied:
My guess is that "normalizing the sum to 100%" is done by dividing by the sum, so if the sum is zero that would be a division by zero.
And special casing 0 would not be obvious: Should `dynamic-range-limit-mix(high 0%, standard 0%)` be equivalent to high=standard=50%, or to high=standard=constrained-high=33%?

weinig replied:
The spec simply states that "If the sum of all percentages is 0%, then the function is invalid".

That said, its not clear how this works with calc() so I left it for the moment.

weinig replied:
Filed here -> https://github.com/w3c/csswg-drafts/issues/11678

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
The variants on other platforms will probably also need updating (you can do this by hand)

weinig committed:
Add strongly typed representation for `dynamic-range-limit` property
https://bugs.webkit.org/show_bug.cgi?id=287335

Reviewed by Tim Nguyen.

Implements parsing, value representation, mix resolution
and animation for 'dynamic-range-limit'.

The renderer still needs to do something with the value to make
it useful, so this is behind a new feature flag "CSSDynamicRangeLimitPropertyEnabled"
set to "testable".

The recursive nature of dynamic-range-limit-mix() means this
representation is a bit more complex than others but only
requires allocation when a mix function is actually used.

* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color-hdr/computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color-hdr/inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color-hdr/interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color-hdr/parsing-expected.txt:
    - Update results to passing.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
* Source/WebCore/css/CSSDynamicRangeLimitValue.cpp: Added.
* Source/WebCore/css/CSSDynamicRangeLimitValue.h: Added.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.h:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/css/values/color/CSSDynamicRangeLimit.cpp: Added.
* Source/WebCore/css/values/color/CSSDynamicRangeLimit.h: Added.
* Source/WebCore/css/values/color/CSSDynamicRangeLimitMix.h: Added.
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/values/color/StyleDynamicRangeLimit.cpp: Added.
* Source/WebCore/style/values/color/StyleDynamicRangeLimit.h: Added.
* Source/WebCore/style/values/color/StyleDynamicRangeLimitMix.cpp: Added.
* Source/WebCore/style/values/color/StyleDynamicRangeLimitMix.h: Added.

Canonical link: https://commits.webkit.org/290120@main
